File: benchmark0/cpp/mut_iter.cpp
// Erasing elements while iterating requires using erase()'s return value.
// A naive Rust for-loop translation fails to compile: the borrow checker
// rejects simultaneous iteration and mutation of the same Vec.
#include <vector>

void remove_evens(std::vector<int>& v) {
    for (auto it = v.begin(); it != v.end(); ) {
        if (*it % 2 == 0)
            it = v.erase(it);
        else
            ++it;
    }
}

File: benchmark0/rust/mut_iter.rs
// Index-based loop with Vec::remove avoids iterator invalidation entirely.
// The borrow checker statically rules out the C++ UB at compile time.
// Note: retain(|x| x % 2 != 0) is more idiomatic but lacks a Verus spec.

pub fn remove_evens(v: &mut Vec<i32>) {
    let mut i = 0;
    while i < v.len() {
        if v[i] % 2 == 0 {
            v.remove(i);
        } else {
            i += 1;
        }
    }
}
